pennylane/vqe/vqe.py

```diff
@@ -129,6 +130,70 @@ def __str__(self):
 
         return "\n+ ".join(terms)
 
+    def __matmul__(self, H):
+
+        coeffs1 = self.coeffs.copy()
+        terms1 = self.ops.copy()
+
+        coeffs2 = H.coeffs
+        terms2 = H.ops
+
+        coeffs = [i[0] * i[1] for i in itertools.product(coeffs1, coeffs2)]
+        term_list = itertools.product(terms1, terms2)
+
+        terms = [qml.operation.Tensor(i[0], i[1]) for i in term_list]
+
+        return qml.Hamiltonian(coeffs, terms)
+
+    def __add__(self, H):
```

Comment:
As far as I understand, the way this is implemented, If we had for example `H_1 = 1*Z + 2*Z` and `H_2 = 3*Z -2*Z`, then `H_1 + H_2= 4*Z`, right? That's great!

But would something similar work when taking the tensor product? So if we had instead `H_1 = 1*Z1 + 2*Z1` and `H_2 = 3*Z2 -2*Z2`, is the output `H_1 @ H_2 = 4*Z1@Z2`?

Reply:
Also, what happens if you add Hamiltonians defined on different wires?

Reply:
To address the first point, no, I believe that like-terms are only combined upon addition or subtraction, but this should be an easy change!

Reply:
As for the second point, when two Hamiltonians with different wires are added, the terms of the Hamiltonians are just combined

Reply:
> `H_1 = 1*Z + 2*Z` and `H_2 = 3*Z -2*Z`, then `H_1 + H_2= 4*Z`

@Lucaman99 not sure how far you have already gotten here, but it would be useful to have a `Hamiltonian.simplify()` method for this. This could be called automatically after initialization, which would be much more 'global':

```pycon
>>> H = qml.Hamiltonian([3, 2], [qml.PauliZ(0), qml.PauliZ(0)])
>>> print(H)
(3) [Z0]
+ (2) [Z0]
```

Reply:
@josh146 This is a fantastic idea!